### PR TITLE
Add global keyboard shortcuts for all tab bar widgets

### PR DESCRIPTION
**Pull request by g547315**
## References
The intended functionality here is to have a keyboard shortcuts for every widget on the tab bar that can be used from anywhere in the application to toggle it open and close

## Code changes

- Added a new commands with the functionality to activate, close and shift tab focus on to the chosen element on tab-bar

- Exposed this command in the command palette

## User-facing changes
New commands to activate:

- Left sidebar widgets with alt + number
- Right sidebar widgets with alt + shift + number

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by g547315** [comment]
Please review @gabalafou

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

File: packages/application-extension/schema/commands.json
Comment (krassowski) [inline packages/application-extension/schema/commands.json:52]: Why this shortcut? Is it standard? Would users know to use it?

File: packages/application-extension/src/index.tsx
Comment (krassowski) [inline packages/application-extension/src/index.tsx:337]: The element should be retrieved programatically from `labShell` - relying from on element IDs is error-prone because we would not notice that something is wrong if the IDs were changed. IDs would be re acceptable if there is _really_ no other way and there is an accompanying test case.

**Review by krassowski — CHANGES_REQUESTED** [review]

File: packages/application-extension/schema/commands.json
Comment (g547315) [reply packages/application-extension/schema/commands.json:52]: This shortcut was chosen due to its proximity to the tab key, secondly for keyboard users there is limited disruption to their hand alignment and positioning when using the shortcut.
This shortcut is not standard as the problem it solves is unique to this application (tab trap in notebooks). It also allows keyboard users to navigate the app more easily as an additional benefit

**Comment by g547315** [comment]
please update documentation snapshots
please update galata snapshots

**Review by krassowski — CHANGES_REQUESTED** [review]
I do not believe that we should introduce such an non-standard shortcut (alt shift z) - this risks conflicts with other applications, extensions, and user actions and is not discoverable.

I think that the command proposed in this PR can be merged if made into generic, e.g. `application.focusSidebarElement` taking arguments `side: 'left' | 'right'` and `index: number`. We can then either to define a shortcut for each item in the sidebar, or just let user configure shortcuts that suit them.

For example, Slack uses ctrl + number and the numbers show up upon pressing ctrl:

image

The problem is that user may prefer to use `ctrl + number` for tab bar, not sidebar and then there are two sidebars - so it may require a combos like `alt + number` for left and `alt + shift + number` for right. Ideally user could modify which combo used for left/right sidebar and which for tabbar (which does not need to be implemented in this RP).

As for focusing **any** specific element in the sidebar, this can already be achieved using existing commands like `filebrowser:toggle-main` (ctrl + shift + f) `toc:show-panel` (ctrl + shift + k), `debugger:show-panel` (ctrl + shift + e), `extensionmanager:show-panel` (extension manager)

**Comment by g547315** [comment]
Thanks for the feedback:

I will make the function more generic accepting varied arguments. Each argument combination will correspond to a keyboard shortcut, facilitating focus on the designated element. For instance, `"args":` `side: 'left'`, `index: 0` will correlate to the shortcut (ctrl shift 1), focusing on the first element of the left tab bar. 

Concerning the display of numbers on elements: using ctrl shift can reveal numbers for the left tab bar, while only ctrl can do so for the right tab bar. It might be best left to be address in a subsequent PR.

@krassowski

**Comment by g547315** [comment]
Please review 
Changes  implemented:

- Refactor to generic method 
- Alt + number for left sidebar and alt + shift + number for right sidebar.

Still to be implemented in another PR:
- Numbers show up upon pressing alt or alt + shift: (Lumino Repo)
- User customization to modify which combo used for left/right sidebar and which for tabbar

File: galata/test/documentation/general.test.ts-snapshots/interface-right-common-documentation-linux.png
Comment (krassowski) [inline galata/test/documentation/general.test.ts-snapshots/interface-right-common-documentation-linux.png:1]: This can be reverted. Apologies about the bot adding flaky snapshots from time to time.

File: packages/application-extension/src/index.tsx
Comment (krassowski) [inline packages/application-extension/src/index.tsx:348]:
```suggestion
          }
        }
```

**Comment by krassowski** [comment]
In the latest commit, https://github.com/jupyterlab/jupyterlab/pull/14799/commits/6c3a552bdc2580813ea82861b2bbb6b21f1a82eb I rewritten the logic to first open the sidebar widget and the focus on it. @g547315 what do you think about this approach? If not, please undo my changes.

However, the focus part does not seem to be working well. It appears to that sidebars do not receive focus in the first place. It maybe be different with https://github.com/jupyterlab/lumino/pull/612. Does this PR depend on it?

**Comment by g547315** [comment]
@Krassowski I think the approach is good, I would note as per your previous comment

There are existing commands like filebrowser:toggle-main (ctrl + shift + f) toc:show-panel (ctrl + shift + k), debugger:show-panel (ctrl + shift + e), extensionmanager:show-panel (extension ctrl + shift + x) to activate these tabs so the might be some redundancy but i think this is ok as this will activate and shift tab focus 

Secondly tab focus depended on this [PR](https://github.com/jupyterlab/lumino/pull/583) which was then changed by the PR you mentioned so Yes

**Comment by g547315** [comment]
update galata snapshots

**Comment by krassowski** [comment]
I am not convinced whether an update is needed here. I re-triggered the regression tests run to see if it passes. For future, "please " prefix is needed for bot to pick up the comment.

**Comment by g547315** [comment]
I think this ready for merging or is there still work to be done

**Comment by krassowski** [comment]
For me it seems good but I would like to have an impartial opinion (since I contributed to this one), so waiting for a review (some maintainers are just coming back from holidays so the queue is longer this time of year).

I mentioned this PR (along other PRs with "needs review" label) during the dev meeting yesterday. If you would like to push it forward, help with reviewing other PRs would be appreciated if you feel like it ([_we all love puppies_](https://github.com/jupyterlab/jupyterlab/blob/main/CORPORATE.md)).

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:344]:
```suggestion
            return;
```
Maybe it can do nothing silently.

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:348]: Shouldn't it be useful to toggle the panel instead of just activating it ?
e.g. toggle-filebrowser command

**Review by brichet — COMMENTED** [review]
Thanks @g547315 and @krassowski, that's really nice.

I have some comments below.

File: packages/application-extension/src/index.tsx
Comment (g547315) [reply packages/application-extension/src/index.tsx:348]: Unsure of the indented change here

Not sure what the difference of Activating the widget (simulating an on click) and toggling the widget.
secondly it is unclear to me that all the tab bar widgets can be toggled and if its even possible to access it by just having the widgets id.

The intended functionality here is to have a keyboard shortcut for every widget on the tab bar that can be used from anywhere in the application to activate and more tab focus to that widget

Comment (brichet) [reply packages/application-extension/src/index.tsx:348]: My idea was to be able to close an open widget.
It might be interesting to be able to close the sidebar using the same keyboard shortcut used to open it.

For example, I'm not sure if there is a shortcut to close the right panel.

Comment (g547315) [reply packages/application-extension/src/index.tsx]: changes made, can you please review

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:368]:
```suggestion
        label: trans.__('Toggle Sidebar Element'),
```

File: packages/application-extension/schema/commands.json
Comment (brichet) [inline packages/application-extension/schema/commands.json:42]: This shortcut is already used to open the 'running kernels' sidepanel:
https://github.com/jupyterlab/jupyterlab/blob/897418f0f25ffbbee3aa06a62770a53246823f85/packages/running-extension/schema/plugin.json#L56-L58

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

**Comment by g547315** [comment]
please update documentation snapshots
please update galata snapshots

**Comment by g547315** [comment]
lease update documentation snapshots
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

**Comment by gabalafou** [comment]
I'm a little confused. Is this ready for review? When I tried testing this PR, the command fails because no arg (left/right) is provided to it when you execute the command from the command palette. In other words, it hits [line 373](https://github.com/jupyterlab/jupyterlab/pull/14799/files#diff-88062eb06c9ca4fb9080028ccdb5d0169b6053e050452fc40c33e767ad5ca0c6R373):

```js
throw Error(`Unsupported sidebar: ${args.side}`);
```

I'm also a little bit lost on what the overall goal of the PR is. I think the PR's description might be out of date. My best guess at what this PR aims to achieve, given the current state of the code is:

1. Provide a command to move browser focus to the left sidebar
2. Provide a command to move browser focus to the right sidebar 

I'm guessing that implies that if the sidebar is closed/collapsed/hidden, the command would first have to open/expand/unhide it. I'm also guessing that we want to put focus not on the tab-button itself but on the panel that clicking on the the tab-button opens/closes.

**Comment by krassowski** [comment]
> when you execute the command from the command palette

This command is not intended to be invoked via command palette - it should not be there, good catch. The implementation that I worked on included shortcuts for each possible sidebar item (see https://github.com/jupyterlab/jupyterlab/pull/14799/commits/a2ba278d5660cd214421a869d7e2c61bc1326148) but it must have been removed later. I did not follow and do not understand why.

**Comment by krassowski** [comment]
It looks like the shortcuts were removed by this commit:

image

https://github.com/jupyterlab/jupyterlab/pull/14799/commits/aa571ac7f7c585d1c48180e04967bb9a11131601

Apologies if I missed it but I do not see relevant discussion around it, nor good explanation in the commit message.

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

**Comment by g547315** [comment]
please update documentation snapshots

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
Apologies I have updated the PR's description and re added the commands
@gabalafou

**Comment by g547315** [comment]
> I'm a little confused. Is this ready for review? When I tried testing this PR, the command fails because no arg (left/right) is provided to it when you execute the command from the command palette. In other words, it hits [line 373](https://github.com/jupyterlab/jupyterlab/pull/14799/files#diff-88062eb06c9ca4fb9080028ccdb5d0169b6053e050452fc40c33e767ad5ca0c6R373):
> 
> ```js
> throw Error(`Unsupported sidebar: ${args.side}`);
> ```
> 
> I'm also a little bit lost on what the overall goal of the PR is. I think the PR's description might be out of date. My best guess at what this PR aims to achieve, given the current state of the code is:
> 
> 1. Provide a command to move browser focus to the left sidebar
> 2. Provide a command to move browser focus to the right sidebar
> 
> I'm guessing that implies that if the sidebar is closed/collapsed/hidden, the command would first have to open/expand/unhide it. I'm also guessing that we want to put focus not on the tab-button itself but on the panel that clicking on the the tab-button opens/closes.

Apologies I have updated the PR's description and re added the commands

**Comment by g547315** [comment]
This PR is ready for re-review @gabalafou gabalafou

**Comment by krassowski** [comment]
Kicking tests

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:108]:
```suggestion
  export const toggleSidebarWidget: string =
    'application:toggle-sidebar-widget';
```

For a better readability, we should rename the command (it should also be updated in *commands.json*).

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:238]:
```suggestion
    function setTabFocus(focusElement: HTMLElement | null) {
      if (focusElement) {
        focusElement.focus();
      }
    }
```
The test can probably be simplified.

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:367]:
```suggestion
      commands.addCommand(CommandIDs.toggleSidebarWidget, {
```

Renaming for readability (see above).

File: packages/application-extension/src/index.tsx
Comment (brichet) [inline packages/application-extension/src/index.tsx:380]:
```suggestion
          const focusElement: HTMLElement | null = document.querySelector(
```

**Review by brichet — CHANGES_REQUESTED** [review]
Thanks @g547315, this looks great.

I have some comments below, mostly for code readability.

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update galata snapshots

File: galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json
Comment (brichet) [inline galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json:35]: This command id should be updated.

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

File: galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json
Comment (g547315) [reply galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json]: Updated

**Comment by brichet** [comment]
Bumping the CI

**Review by brichet — APPROVED** [review]
Thanks @g547315, it looks good to me.

File: galata/test/jupyterlab/readonly.test.ts-snapshots/readonly-jupyterlab-linux.png
Comment (brichet) [inline galata/test/jupyterlab/readonly.test.ts-snapshots/readonly-jupyterlab-linux.png:1]: This snapshot should probably be restored

**Comment by g547315** [comment]
please update galata snapshots

File: galata/test/jupyterlab/readonly.test.ts-snapshots/readonly-jupyterlab-linux.png
Comment (g547315) [reply galata/test/jupyterlab/readonly.test.ts-snapshots/readonly-jupyterlab-linux.png:1]: snapshot restored

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update galata snapshots

**Comment by g547315** [comment]
please update documentation snapshots

**Comment by g547315** [comment]
@gabalafou
@krassowski 

Can this PR be merged?, The failing test are not due to the code changes make

File: galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png
Comment (krassowski) [inline galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png:1]: This change appears to be causing a test failure:

>    [documentation] › test/documentation/general.test.ts:19:7 › General › Welcome ──────────────────

Can you revert this snapshot update manually?

Comment (brichet) [reply galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png:1]: I think @krassowski means you should probably revert it using git (locally) then push your changes, instead of using the galata comment in github.

Comment (g547315) [reply galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png:1]: Thanks for the clarification @brichet. @krassowski 
I have now updated the snap shot in question manually

**Comment by brichet** [comment]
@krassowski what do you think about merging this PR.
The failing tests are not related.

File: packages/application-extension/src/index.tsx
Comment (krassowski) [inline packages/application-extension/src/index.tsx:367]: This is not the case I think? By default it errors as `args.index` is `undefined`?

```suggestion
```

Comment (krassowski) [reply packages/application-extension/src/index.tsx:367]: Also, caption may not be the best place to mention it even if it were working.

Comment (krassowski) [reply packages/application-extension/src/index.tsx:367]: @brichet good to merge once this thread is addressed. Feel free to go either way.

File: galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json
Comment (brichet) [inline galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json:37]:
```suggestion
```
This one should also be removed, then we should be good to merge.

File: packages/application-extension/src/index.tsx
Comment (g547315) [reply packages/application-extension/src/index.tsx]: I have removed the caption

File: galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json
Comment (g547315) [reply galata/test/documentation/commands.test.ts-snapshots/commandsList-documentation-linux.json]: Removed

**Pull request merged**
